File: tea/src/main/java/turtleduck/tea/Shell.java
package turtleduck.tea;

import static turtleduck.tea.Diagnostics.*;
import static turtleduck.tea.HTMLUtil.*;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.Collections;

import org.slf4j.Logger;
import org.teavm.jso.dom.html.HTMLElement;

import turtleduck.colors.Color;
import turtleduck.colors.Colors;
import turtleduck.messaging.Connection;
import turtleduck.messaging.Message;
import turtleduck.messaging.Reply;
import turtleduck.messaging.Router;
import turtleduck.messaging.ShellService;
import turtleduck.messaging.generated.ShellServiceProxy;
import turtleduck.text.Location;
import turtleduck.util.Array;
import turtleduck.util.Dict;
import turtleduck.util.JsonUtil;
import turtleduck.util.Logging;

public class Shell {
    public static final Logger logger = Logging.getLogger(Shell.class);
    private int lineNum = 0;
    public static final Color VARCOLOR = Colors.OLIVE;
    public static final Color TYPECOLOR = Colors.LIGHT_BLUE;
    public static final Color TEXTCOLOR = Colors.LIME;
    private ShellService service;
    private Connection conn;
    private Language language;

    Shell(Language lang, Connection conn) {
        this.language = lang;
        this.service = conn != null ? new ShellServiceProxy(conn.id(), conn.router()) : null;
        this.conn = conn;
    }

    Shell(Language lang, ShellService service, Connection conn) {
        this.language = lang;
        this.service = service;
        this.conn = conn;
    }

    public Connection connection() {
        return conn;
    }

    public Language language() {
        return language;
    }

    public String shellName() {
        return language.shellName;
    }

    public String fileExtension() {
        return Languages.langToExt(language.id, true);
    }

    public void printExceptions(Dict msg, LanguageConsole console) {
        Dict ex = msg.get(ShellService.EXCEPTION);
        HTMLElement output = console.outputElement();
        if (ex != null) {
            if (output != null) {
                HTMLElement trace = element("details", clazz("traceback fold"));
                boolean first = true;
                for (String s : ex.get(Reply.TRACEBACK).toListOf(String.class)) {
                    logger.info("trace: " + ex);
                    HTMLElement li;
                    if (first) {
                        li = element("summary", s);
                        // JSUtil.activateToggle(li, "open", trace);
                        first = false;
                    } else {
                        li = element("li", s);
                    }
                    trace.appendChild(li);
                }
                HTMLElement elt = div(clazz("diag diag-error"), //
                        element("p", clazz("diag-header"), //
                                span(ex.get(Reply.ENAME), clazz("exception-name")), ": ", //
                                span(ex.get(Reply.EVALUE), clazz("exception-message"))), //
                        trace);
                output.appendChild(elt);

            } else {
                console.println(ex.get(Reply.ENAME) + ": " + ex.get(Reply.EVALUE), colorOf("error"));
                for (String trace : ex.get(Reply.TRACEBACK).toListOf(String.class)) {
                    console.println(trace, colorOf("error"));
                }
            }
        }
    }

    public Array printDiags(Dict msg, LanguageConsole console) {
        Array diags = msg.get(ShellService.DIAG);
        String worstLevel = "none";
        HTMLElement output = console.outputElement();
        for (Dict diag : diags.toListOf(Dict.class)) {
            String name = diag.get(Reply.ENAME);
            String level = levelOf(name);
            worstLevel = worstOf(worstLevel, level);
            Location loc = Location.fromString(diag.get(ShellService.LOC));
            if (output != null) {
                String code = msg.get(ShellService.CODE);
                HTMLElement codefrag = null;
                if (code != null) {
                    String before = loc.before(code).replaceAll("(?m)^.*\n", "");
                    String after = loc.after(code).replaceAll("(?m)\n.*$", "");

                    if (loc.length() > 0)
                        codefrag = element("p", clazz("diag-code"), span(before),
                                span(loc.substring(code), clazz("diag-" + level)), span(after));
                    else
                        codefrag = element("p", clazz("diag-code"), span(before),
                                span(clazz("diag-between diag-" + level)), span(after));
                }
                HTMLElement elt = div(clazz("diag diag-" + levelOf(name)), //
                        element("p", clazz("diag-header"), name + " at ", //
                                element("a", loc.toString(), attr("href", loc.path()), attr("target", "_blank")), ":"),
                        codefrag, //
                        element("p", clazz("diag-message"), diag.get(Reply.EVALUE)));

                output.appendChild(elt);
            } else {
                console.println(name + " at " + loc, colorOf(levelOf(name)));
                console.println(diag.get(Reply.EVALUE), colorOf(levelOf(name)));
            }
            if (console.hasDiagnostics()) {
                try {
                    URI uri = new URI(diag.get(ShellService.LOC));
                    Location l = new Location(uri);
                    diag.put("level", level);
                    console.diagnostic(diag, l);

                    // addAnno(state, l.start(), l.length(), "error",
                    // diag.get(Reply.ENAME) + ": " + diag.get(Reply.EVALUE));
                } catch (URISyntaxException ex) {
                    Browser.addError(ex);
                }
            }
        }
        if (output != null) {
            output.setClassName(output.getClassName() + " with-diag with-diag-" + worstLevel);
        }
        return diags;
    }

    public boolean specialCommand(String line, LanguageConsole console) {
        String[] split = line.split("\\s+", 2);
        String cmd = split[0];
        String args = split.length == 2 ? split[1].trim() : "";
        if (cmd.equals("/:ls")) {
            console.promptBusy();
            Client.client.fileSystem.readdir(args).onComplete(files -> {
                int max = files.stream().mapToInt(str -> str.length()).max().orElse(0);
                int width = 80; // TODO terminal.getCols();
                int cols = 1;
                logger.info("0: max={}, width={}, cols={}, files={}", max, width, cols, files);
                if (max > 0 && max < width) {
                    cols = Math.min(files.size(), width / (max + 1));
                    max = width / cols;
                }
                logger.info("1: max={}, width={}, cols={}", max, width, cols);
                int c = 0;
                Collections.sort(files);
                for (String str : files) {
                    console.print(String.format("%-" + max + "s", str));
                    if (++c >= cols) {
                        console.println();
                        c = 0;
                    }
                }
                if (c != 0)
                    console.println();
                console.promptNormal();
            }, err -> {
                logger.error("error: {}", JsonUtil.encode(err));
                printError("", err, console);
            });

            return true;
        } else if (cmd.equals("/:mkdir")) {
            console.promptBusy();
            Client.client.fileSystem.mkdir(args).onComplete(res -> {
                console.promptNormal();
            }, err -> {
                logger.error("error: {}", JsonUtil.encode(err));
                printError("", err, console);
            });

            return true;
        } else if (cmd.equals("/:cd")) {
            console.promptBusy();
            Client.client.fileSystem.chdir(args).onComplete(res -> {
                console.println("cd " + res);
                console.promptNormal();
            }, err -> {
                logger.error("error: {}", JsonUtil.encode(err));
                printError("", err, console);
            });

            return true;
        } else if (cmd.equals("/:readbin")) {
            console.promptBusy();
            Client.client.fileSystem.readbinfile(args).onComplete(res -> {
                logger.info("readbinfile: {}", res);
                String s = "";
                for (int i = 0; i < res.length; i++) {
                    s = s + String.format("%02x ", res[i]);
                    if (i > 0 && i % 40 == 0) {
                        console.println(s);
                        s = "";
                    }
                }
                console.println(s);
                console.promptNormal();
            }, err -> {
                logger.error("error: {}", JsonUtil.encode(err));
                printError("", err, console);
            });

            return true;
        } else if (cmd.equals("/:readtext")) {
            console.promptBusy();
            Client.client.fileSystem.readtextfile(args).onComplete(res -> {
                logger.info("readtextfile: {}", res);
                console.println(res);
                console.promptNormal();
            }, err -> {
                logger.error("error: {}", JsonUtil.encode(err));
                printError("", err, console);
            });

            return true;
        } else if (cmd.equals("/:writebin")) {
            console.promptBusy();
            byte[] data = { (byte) 0xde, (byte) 0xad, (byte) 0xbe, (byte) 0xef };
            Client.client.fileSystem.writebinfile(args, data).onComplete(res -> {
                logger.info("writebinfile: {}", data);
                String s = "";
                for (int i = 0; i < data.length; i++) {
                    s = s + String.format("%02x ", data[i]);
                    if (i > 0 && i % 40 == 0) {
                        console.println(s);
                        s = "";
                    }
                }
                console.println(s);
                console.promptNormal();
            }, err -> {
                logger.error("error: {}", JsonUtil.encode(err));
                printError("", err, console);
            });

            return true;
        } else if (cmd.equals("/:writetext")) {
            console.promptBusy();
            Client.client.fileSystem.writetextfile(args, "deadbeef").onComplete(res -> {
                console.promptNormal();
            }, err -> {
                logger.error("error: {}", JsonUtil.encode(err));
                printError("", err, console);
            });

            return true;
        } else if (cmd.equals("/load_language")) {
            Client.client.loadLanguage(args);
            console.promptNormal();
            return true;
        } else if (cmd.equals("/open")) {
            String l = Languages.extToLang(args);
            if (l.isEmpty())
                l = language.id;
            Client.client.editorImpl.open(args, null, l);
            Client.client.editorImpl.focus();
            console.promptNormal();
            return true;
        } else if (cmd.equals("//router_local")) {
            console.print(Client.client.router.command(args));
            console.promptNormal();
            return true;
        } else if (cmd.equals("/:echo")) {
            console.println(args);
            console.promptNormal();
            return true;
        } else if (cmd.equals("//router_remote")) {
            if (conn != null) {
                Message msg = Message.writeTo("$remote", "$router").putContent(Router.COMMAND, args).done();
                conn.send(msg).onSuccess(result -> {
                    console.print(result.get(Router.RESULT));
                });
                console.promptNormal();
            }
            return true;
        } else if (cmd.equals("//send")) {
            if (conn != null) {
                String msg_type = args;
                int i = args.indexOf(' ');
                Dict content;
                if (i >= 0) {
                    msg_type = args.substring(0, i);
                    args = args.substring(i);
                    content = JSUtil.decodeDict(args);
                } else {
                    content = Dict.create();
                }
                Message msg = Message.writeTo("$remote", msg_type).content(content).done();
                conn.send(msg).onSuccess(result -> {
                    console.print(result.get(Router.RESULT));
                });
                console.promptNormal();
            }
            return true;
        }
        return false;
    }

    public boolean processResult(Dict msg, boolean processIncomplete, LanguageConsole console) {
        logger.info("exec result: {}", msg);
        HTMLElement output = console.outputElement();
        logger.info("output elt: {}", output);
        logger.info("complete: {}", msg.get(ShellService.COMPLETE));
        logger.info("multi: {}", msg.get(ShellService.MULTI));
        logger.info("multi-s: {}", msg.getArray("multi"));
        if (msg.get(ShellService.COMPLETE) || processIncomplete) {
            for (Dict result : msg.get(ShellService.MULTI).toListOf(Dict.class)) {
                printDiags(result, console);
                printExceptions(result, console);
                String value = result.get(ShellService.VALUE);
                String name = result.get(ShellService.NAME);
                String type = result.get(ShellService.TYPE);
                Dict display = result.get("display", Dict.create());
                logger.info("result: {}, {}, {}", value, name, type);
                if (value != null) {
                    if (output != null) {
                        HTMLElement elt = span(clazz("eval-result"));
                        if (type != null) {
                            elt.appendChild(span(type, clazz("cmt-typeName")));
                            elt.appendChild(text(" "));
                        }
                        if (name != null) {
                            elt.appendChild(span(name, clazz("cmt-variableName")));
                            elt.appendChild(text(" = "));
                        }
                        elt.appendChild(span(value, clazz("cmt-literal")));
                        if ("img".equals(display.getString("display"))) {
                            ArrayView<?> data = display.get("data", ArrayView.class);
                            String url = display.getString("url");
                            if (data != null) {
                                Browser.consoleLog("data", data.get());
                                url = JSUtil.createObjectURL(data.get(), display.get("format", "image/png"));
                                logger.info("image: {}", url);
                            }

                            if (url != null && url.startsWith("blob:")) {
                                elt.appendChild(element("img", attr("src", url)));
                                // if(data != null)
                                // JSUtil.revokeObjectURL(url);
                            }
                        }
                        output.appendChild(elt);
                    } else {
                        String v = TEXTCOLOR.applyFg(value) + "\n";
                        if (name != null) {
                            v = VARCOLOR.applyFg(name) + " = " + v;
                        }
                        if (type != null) {
                            v = TYPECOLOR.applyFg(type) + " " + v;
                        }
                        console.print(v);
                    }
                }
            }
            Client.client.showHeap(msg);
            Client.client.userlog("Eval: done");
            return true;
        } else {
            Client.client.userlog("Eval: incomplete input");
            if (output != null) {
                output.appendChild(span("…", clazz("diag-error")));
            } else {
                console.println("…", Colors.MAROON);
            }
            // String code = msg.get(ShellService.CODE);
            // terminal.paste(code);
            // if (code.endsWith(";"))
            // readline.keyHandler(KeyEvent.create(KeyCodes.Navigation.ARROW_LEFT, "", 0,
            // 0));
            return false;
        }
    }

    public void evalLine(String line, LanguageConsole console) {
        evalLine(line, lineNum++, null, console);
    }

    public void evalLine(String line, int id, Dict opts, LanguageConsole console) {
        if (opts == null)
            opts = Dict.create();
        Client.client.userlog("Eval: running...");

        service.eval(line, id, opts)//
                .onSuccess(msg -> {
                    Client.client.userlog("Eval: finished");
                    if (processResult(msg, false, console)) {
                        console.promptNormal();
                    } else {
                        console.promptMore(line);
                    }
                })//
                .onFailure(msg -> {
                    Client.client.userlog("Eval: internal error");
                    logger.info("exec error: " + msg);
                    printError("INTERNAL ERROR: ", msg, console);

                });
        // }
    }

    public void printError(String prefix, Dict msg, LanguageConsole console) {
        String ename = msg.get(Reply.ENAME);
        String evalue = msg.get(Reply.EVALUE, null);
        Array trace = msg.get(Reply.TRACEBACK);
        console.println(prefix + ename + (evalue != null ? (" : " + evalue) : ""), Colors.RED);
        for (String frame : trace.toListOf(String.class)) {
            console.println(frame, Colors.MAROON);
        }
        console.promptNormal();

    }

    public ShellService service() {
        return service;
    }

}

File: tea/src/main/webroot/js/Shell.ts
import { Hole, html } from 'uhtml';
import { BorbTerminal, DisplayData } from '../borb/Terminal';
import {
    Messaging,
    BaseConnection,
    Payload,
    Method,
    Message,
    Connection,
} from '../borb/Messaging';
import { Language } from './Language';
import Settings from '../borb/Settings';
import { LineEditor } from '../borb/LineEditor';
import { turtleduck } from './TurtleDuck';
import { uniqueId } from '../borb/Common';
import { Optionals } from './Util';
import { defaultsDeep } from 'lodash-es';
import { Location } from './Location';
interface ShellMessage {
    prompt?: string;

    doc?: string;

    /**
     * Reply: True if source code is complete, false if more input is needed
     */
    complete: boolean; // = true
    persistent?: boolean;
    active?: boolean;

    names?: string[];
    docs?: string[];
    /**
     * Reply: The full name of symbol that was (re)declared/defined, if any.
     *
     * Includes type and parameters
     *
     * @see #DEF
     */
    fullname?: string;

    /**
     * Reply: An array of multiple eval replies, if the input code was split into
     * multiple snippets
     */
    multi: EvalReply[];

    text?: string;
    heapUse?: number;
    heapTotal?: number;
    heapMax?: number;
    cpuTime?: number; // = 0.0
}
interface Diag {
    msg: string;
    line?: string;
    start: number;
    end: number;
    pos: number;
    ename: string;
    evalue: string;
    loc: string;
}
interface Exception {
    exception: string;
    ename: string;
    evalue: string;
    traceback: string[];
    cause: Exception;
}
interface ShellService {
    eval(msg: EvalRequest): Promise<EvalReply>;
    refresh(msg: UpdateRequest): Promise<{}>;
}
interface ExplorerService {
    update(msg: UpdateRequest): Promise<{}>;
}

interface TerminalService {
    prompt(msg: { prompt: string; language: string }): Promise<{}>;

    print: Method<{ text: string; stream: string }, {}>;

    display: Method<{ data: DisplayData; stream: string }, {}>;

    read_request: Method<
        {
            prompt: string;
            language?: string;
        },
        { text: string }
    >;
}
/*
interface TerminalService {
    prompt(msg: { prompt: string; language: string }): Promise<{}>;

    print(msg: { text: string; stream: string }): Promise<{}>;

    display(msg: { data: DisplayData; stream: string }): Promise<{}>;

    read_request(msg: {
        prompt: string;
        language?: string;
    }): Promise<{ text: string }>;
}
*/

export interface EvalRequest extends Payload {
    /**
     * Request/(Reply): Code to be executed. In a reply, this is the code that was
     * actually executed, which may be different (e.g, minor syntax corrections,
     * such as missing semicolon)
     */
    code: string;
    /**
     * Request/Reply: A numeric reference provided by the caller
     */
    ref: number;
    /**
     * Request: Additional options to the evaluator
     */
    opts?: Payload;
}

const defaultEvalRequest: Optionals<EvalRequest> = {
    opts: {},
};

export interface SnipInfo {
    /**
     * Reply: The kind of snippet that was evaluated.
     *
     * One of error, expression, import, method, statement, type or var; with
     * optional subtypes provided after a dot. (E.g., <code>var.decl.init</code>)
     */
    snipkind: string;
    /**
     * Reply: Identifier for an evaluated snippet; this will match the id sent in
     * any Explorer updates
     */
    snipid: string;
    snipns: string;
}
export interface EvalResult {
    snippet?: SnipInfo;
    /**
     * Reply: An exception, if one was thrown.
     *
     * Includes "exception" (exception class name), "message" (the message), "trace"
     * (array with stack trace), and optional "cause" (another exception)
     */
    exception?: Exception;
    /**
     * Reply: An array of error/diagnostic messages.
     *
     * Message fields include "msg", "start", "end", "pos"
     */
    diag: Diag[];
    /**
     * Request/(Reply): Code to be executed. In a reply, this is the code that was
     * actually executed, which may be different (e.g, minor syntax corrections,
     * such as missing semicolon)
     */
    code: string;
    /**
     * Request/Reply: The code location
     */
    loc?: string;
    /**
     * Reply: True if evaluation involved executing code (i.e., not just declaring
     * something)
     */
    exec?: boolean;
    /**
     * Reply: True if evaluation involved defining/declaring something
     *
     * @see #SYMBOL
     */
    def?: boolean;
    /**
     * Reply: The symbol that was (re)declared/defined, if any.
     *
     * @see #DEF
     */
    name?: string;
    signature?: string;
    /**
     * Reply:
     */
    value?: unknown;
    /**
     * Reply: The type of the result, if any.
     *
     * @see #VALUE
     */
    type?: string;

    /**
     * Icon for the result type, if any
     */
    icon?: string;
    display?: DisplayData;
}

export interface EvalReply {
    /**
     * Request/Reply: A numeric reference provided by the caller
     */
    ref: number;
    status: 'ok' | 'error' | 'incomplete';
    results: EvalResult[];
}

export interface UpdateRequest {
    info: {
        sym: string;
        signature: string;
        type?: string;
        category: 'method' | 'import' | 'type' | 'var';
        verb: 'created' | 'deleted' | 'updated';
        snippet?: SnipInfo;
    };
}
export class ShellConnection extends BaseConnection {
    constructor(
        router: typeof Messaging,
        public shell: Shell,
        id: string,
        exId: string,
    ) {
        super(router, id, exId);
    }

    deliverRemote(msg: Message, transfers: Transferable[]): void {
        const m = this.shell[msg.header.msg_type] as (
            msg: Payload,
        ) => Promise<Payload>;
        console.log('got request', msg, transfers, m);
        if (typeof m === 'function') {
            m(msg.content).then((reply) => {
                this.deliverHost(this.router.reply(msg, reply));
            });
        }
    }
}

export class Shell implements ShellService, TerminalService, ExplorerService {
    id: string;
    terminal: BorbTerminal;
    language: Language;
    conn: Connection;
    languageName: string;
    shellName: string;
    lineEditor: LineEditor;

    constructor(language: Language) {
        this.language = language;
    }

    async init(language: Language, id: string) {
        if (!language || !id) throw new Error('Illegal arguments');
        this.id = `${id}_shell`;
        if (!this.terminal) this.terminal = new BorbTerminal();
        this.terminal.setAttribute('shell', language.shellName);
        this.terminal.setAttribute('frame-title', language.shellTitle);
        this.terminal.setAttribute('language', language.name);
        this.terminal.setAttribute('icon', language.icon);
        this.languageName = language.name;
        this.shellName = language.name;
        if (!this.terminal.id) this.terminal.id = `${id}_terminal`;
        this.conn = new ShellConnection(
            Messaging,
            this,
            this.id,
            `${id}_explorer`,
        );
        this.lineEditor = new LineEditor(this.terminal, this.language);
        this.lineEditor.onEnter = this.onEnter;
        console.log('waiting for line editor...');
        return Promise.resolve();
    }
    mountTerminal(parent?: HTMLElement) {
        if (parent) {
            parent.appendChild(this.terminal);
            return;
        }
        const defaultFrame = document.getElementById(
            Settings.getConfig('terminal.defaultFrame', 'terminal'),
        );
        if (defaultFrame) {
            defaultFrame.appendChild(this.terminal);
            return;
        }
        const existingTerminal = document.querySelector(
            'borb-terminal',
        ) as HTMLElement;
        if (existingTerminal) {
            existingTerminal.insertAdjacentElement('afterend', this.terminal);
            return;
        }
    }
    onEnter = async (
        line: string,
        lineId: number,
        outputElement: HTMLElement,
        lineEditor: LineEditor,
    ) => {
        const payload = { code: line, ref: `${lineId}`, opts: {} };
        const result = await Messaging.send(
            payload,
            'eval_request',
            this.language.connectionId,
        );
        if (processResult(result as unknown as EvalReply, this.terminal, false))
            return {};
        else return { more: line };
    };
    async prompt(msg: { prompt: string; language: string }): Promise<{}> {
        console.error('Method not implemented.');
        return {};
    }
    print: Method<{ text: string; stream: string }, {}> = async ({
        text,
        stream,
    }) => {
        this.terminal.print(text);
        return {};
    };
    printElement: Method<{ elt: HTMLElement; stream: string }, {}> = async ({
        elt,
        stream,
    }) => {
        this.terminal.printElement(elt);
        return {};
    };
    display: Method<{ data: DisplayData; stream: string }, {}> = async ({
        data,
        stream,
    }) => {
        this.terminal.display(data);
        return {};
    };
    read_request: Method<
        { prompt: string; language?: string },
        { text: string }
    > = async ({ prompt, language = '' }) => {
        return { text: 'not implemented' };
    };
    eval = async ({ code, ref, opts }: EvalRequest): Promise<EvalReply> => {
        throw new Error('Method eval_request not implemented.');
    };
    refresh = async ({ info }): Promise<{}> => {
        throw new Error('Method refresh not implemented.');
    };

    update = async ({ info }: UpdateRequest): Promise<{}> => {
        try {
            const defaults: Optionals<typeof info> = {
                type: '',
                snippet: {
                    snipid: info.sym,
                    snipns: 'main',
                    snipkind: info.category,
                },
            };
            const { snippet, sym, verb, type, signature, category } =
                defaultsDeep(info, defaults);
            // if (persistent !== undefined && !persistent) return;
            const sig = signature;
            if (sig && !sig.includes('$')) {
                if (snippet.snipns === 'main') {
                    const s = `${sym} ${verb}`;
                    const div = html.node`<div><a class="prompt">[${
                        snippet.snipid
                    }]</a> ${sym} ${verb} <span class="cmt-keyword">${category}</span> <span class="cmt-variableName" data-snipid=${
                        snippet.snipid
                    }>${sig}</span> ${
                        type
                            ? html`: <span class="cmt-typeName">${type}</span>`
                            : ''
                    }</div>`;
                    this.terminal.printElement(div);
                }
            }
        } catch (e) {}
        return {};
    };
}

function printExceptions(msg: EvalResult) {
    const ex = msg.exception;
    if (ex) {
        const ename = ex.ename || 'Error';
        const evalue = ex.evalue || '';
        const trace = ex.traceback || [];
        const tb = html`<details class="traceback fold">${trace.map((s, i) =>
            i == 0 ? html`<summary>${s}</summary>` : html`<li>${s}</li>`,
        )}</details`;
        return html`<div class="diag diag-error">
            <p class="diag-header">
                <span class="exception-name">${ename}</span>:
                <span class="exception-message">${evalue}</span>
            </p>
            ${tb}
        </div>`;
    } else {
        return '';
    }
}

export const DIAG_LEVELS = ['error', 'warning', 'info', 'debug'];
export function levelOf(name: string) {
    if (DIAG_LEVELS.includes(name)) return name;
    else return 'error';
}
export function worstOf(lvl1: string, lvl2: string) {
    const l1 = DIAG_LEVELS.indexOf(lvl1);
    const l2 = DIAG_LEVELS.indexOf(lvl2);
    if (l1 < 0) return lvl2;
    else if (l2 < 0) return lvl1;
    else if (l1 >= l2) return lvl1;
    else return lvl2;
}
function printDiags(msg: EvalResult) {
    let worstLevel = 'none';
    return (msg.diag || []).map((diag) => {
        const name = diag.ename;
        const level = levelOf(name);
        worstLevel = worstOf(worstLevel, level);
        const loc = Location.fromString(diag.loc);
        let codefrag: Hole;
        let locfrag: Hole;
        const code = msg.code;
        if (loc && code) {
            const [before, err, after] = loc.splitLine(code, diag.line);
            if (err.length > 0) {
                codefrag = html`<span class="diag-before">${before}</span
                    ><span class=${'diag-' + level}>${err}</span
                    ><span> class="diag-after"${after}</span>`;
            } else {
                codefrag = html`<span class="diag-before">${before}</span
                    ><span class=${'diag-between diag-' + level}></span
                    ><span class="diag-after">${after}</span>`;
            }
            locfrag = html`<a href=${loc.toString()}>${loc.toString()}</a>`;
        }
        const elt = html`<div class=${'diag diag-' + level}>
            <p class="diag-header">${name} at ${locfrag || ''}:</p>
            <p class="diag-code">${codefrag || ''}</p>
            <p class="diag-message">${diag.evalue}</p>
        </div>`;
        return elt;
    });
}
/*
    public Array printDiags(Dict msg, LanguageConsole console) {
        Array diags = msg.get(ShellService.DIAG);
        String worstLevel = "none";
        HTMLElement output = console.outputElement();
        for (Dict diag : diags.toListOf(Dict.class)) {
            String name = diag.get(Reply.ENAME);
            String level = levelOf(name);
            worstLevel = worstOf(worstLevel, level);
            Location loc = Location.fromString(diag.get(ShellService.LOC));
            if (output != null) {
                String code = msg.get(ShellService.CODE);
                HTMLElement codefrag = null;
                if (code != null) {
                    String before = loc.before(code).replaceAll("(?m)^.*\n", "");
                    String after = loc.after(code).replaceAll("(?m)\n.*$", "");

                    if (loc.length() > 0)
                        codefrag = element("p", clazz("diag-code"), span(before),
                                span(loc.substring(code), clazz("diag-" + level)), span(after));
                    else
                        codefrag = element("p", clazz("diag-code"), span(before),
                                span(clazz("diag-between diag-" + level)), span(after));
                }
                HTMLElement elt = div(clazz("diag diag-" + levelOf(name)), //
                        element("p", clazz("diag-header"), name + " at ", //
                                element("a", loc.toString(), attr("href", loc.path()), attr("target", "_blank")), ":"),
                        codefrag, //
                        element("p", clazz("diag-message"), diag.get(Reply.EVALUE)));

                output.appendChild(elt);
            } else {
                console.println(name + " at " + loc, colorOf(levelOf(name)));
                console.println(diag.get(Reply.EVALUE), colorOf(levelOf(name)));
            }
            if (console.hasDiagnostics()) {
                try {
                    URI uri = new URI(diag.get(ShellService.LOC));
                    Location l = new Location(uri);
                    diag.put("level", level);
                    console.diagnostic(diag, l);

                    // addAnno(state, l.start(), l.length(), "error",
                    // diag.get(Reply.ENAME) + ": " + diag.get(Reply.EVALUE));
                } catch (URISyntaxException ex) {
                    Browser.addError(ex);
                }
            }
        }
        if (output != null) {
            output.setClassName(output.getClassName() + " with-diag with-diag-" + worstLevel);
        }
        return diags;
    }

*/

function processResult(
    msg: EvalReply,
    terminal: BorbTerminal,
    processIncomplete = false,
) {
    if (msg.status === 'ok' || msg['complete'] || processIncomplete) {
        (msg.results || []).forEach((result, index) => {
            const diags = printDiags(result);
            const ex = printExceptions(result);
            const value = result.value;
            const name = result.name
                ? html`<span class="cmt-variableName">${result.name}</span> = `
                : '';
            const type = result.type
                ? html`<span class="cmt-typeName">${result.type} </span>`
                : '';
            const display = result.display;
            console.log('result', value, name, type);
            // if (value) {
            terminal.printElement(
                html.node`${diags}${ex}<span class="eval-result">${type}${name}<span class="cmt-literal">${result.value}</span>`,
            );
            // }
        });
        // TODO turtleduck.showHeap(msg);
        turtleduck.userlog('Eval: done');
        return true;
    } else {
        turtleduck.userlog('Eval: incomplete input');
        terminal.printElement(html.node`<span class="diag-error">…</span>`);
        return false;
    }
}
